### DISCUSSION TITLE
How to call scaleVirtualMachine api with details parameter

### DISCUSSION THREAD

**Discussion by YashasAnand** [apache/cloudstack#9961] [Q&A] [answered]
I am trying to call scaleVirtualMachine API using python, but the documentation is very vague on how to use details parameter
[https://cloudstack.apache.org/api/apidocs-4.19/apis/scaleVirtualMachine.html](url)

I am trying to create request like this, is this wrong? can anyone pls suggest what is right way to call this api?

request = {
    'command': 'scaleVirtualMachine',  
    'id': 'xys',  # ID of the system VM
    'serviceofferingid': 'xyz', 
    'response': 'json', 
    'details' : [
    {"name": "memory", "value": "3072"},
    {"name": "cpunumber", "value": "2"}
    ]
}

I get error 

{
    "scalevirtualmachineresponse": {
        "uuidList": [],
        "errorcode": 431,
        "cserrorcode": 9999,
        "errortext": "Unable to execute API command scalevirtualmachine due to invalid value [{\"name\": \"memory\", \"value\": \"3072\"}, {\"name\": \"cpunumber\", \"value\": \"2\"}] for parameter details"
    }
}

**Comment by weizhouapache** [comment]
Can you try

["memory": "3072", "cpunumber": "2"]

or similar ?

**Comment by bernardodemarco** [comment]
Hello, @YashasAnand

The `details` parameter of the `scaleVirtualMachine` API is a `MAP`. Basically, the `MAP` parameter type in CloudStack expects the keys and values as:

```bash
details[0].key=value
``` 

So, for example, a VM with 512 MB of RAM and 1 vCPU can be scaled to 1024 MB of RAM and 2 vCPUs by executing the following API call through CloudMonkey:

```bash
scale virtualmachine id=<vm-id> serviceofferingid=<compute-offering-id> details[0].memory=1024 details[0].cpuNumber=2
```

How does your Python script sends the request? Does each key in the `request` dict represent a query param in the HTTP request? If so, then one solution would be to define the details parameter to be something like:

```python
request = {
    'command': 'scaleVirtualMachine',
    'id': '<virtual-machine-id>',
    'serviceofferingid': '<compute-offering-id>',
    'response': 'json',
    'details[0].memory': 1024 ,
    'details[0].cpuNumber': 2
}
```

**Comment by YashasAnand** [comment]
Hi @bernardodemarco @weizhouapache Thank you for your respective responses. Below is my python code. 


    def generate_signature(baseurl, request, secretkey):
    
       encoded_request = {urllib.parse.quote_plus(k): urllib.parse.quote_plus(str(v)) for k, v in request.items()} 
       sorted_request = sorted(encoded_request.items(), key=lambda x: x[0].lower())
       sig_str = '&'.join(['='.join([k.lower(), v.lower()]) for k, v in sorted_request])
    
       sig = hmac.new(secretkey.encode('utf-8'), sig_str.encode('utf-8'), hashlib.sha1).digest()
    
       encoded_sig = base64.b64encode(sig).decode('utf-8')
       final_signature = urllib.parse.quote_plus(encoded_sig)

       request_str = '&'.join(['='.join([k, v]) for k, v in encoded_request.items()])
       return final_signature, request_str
   
    baseurl = 'basrurl'

    request = {
    'command': 'scaleVirtualMachine',  
    'id': 'id',  # ID of the system VM
    'serviceofferingid': 'serviceofferingid', 
    'response': 'json', 
    'details[0].memory': 3072,
    'details[0].cpuNumber': 2
    }
    secretkey = 'secret key'
    signature, request_str = generate_signature(baseurl, request, secretkey)
    req = baseurl + request_str + '&signature=' + signature
    print("Final Request URL:", req)


I am getting response like this but i dont observe any scaling happening on the vm even though i have domain admin access. Pls can you suggest what might be the issue.

`{
    "scalevirtualmachineresponse": {
        "jobid": "f75d1e48-132e-4b5c-9241-9115271dc08e"
    }
}`

**Comment by bernardodemarco** [comment, marked as answer]
APIs in CloudStack can be synchronous/blocking, or asynchronous/non-blocking. Async APIs return a `jobid`, which can be used to poll the result of the API call.

Since the `scaleVirtualMachine` API is asynchronous, the `jobid` was returned, as expected. It means that the request was successfully authenticated and its response can be retrieved using the `queryAsyncJobResult` API:

```bash
query asyncjobresult jobid="<job-id>"
```

When retrieving the result of a job related to the `scaleVirtualMachine` API, a successful response will be something like:

```bash
(localcloud) 🐱 > query asyncjobresult jobid=75f6ebd3-e4ed-477c-b6ff-2577e2e56c30
{
  (...)
  "jobresult": {
    "virtualmachine": {
      "account": "admin",
      "affinitygroup": [],
      "cpunumber": 1,
      "cpuspeed": 512,
      "created": "2024-11-21T15:07:09+0000",
      "deleteprotection": false,
      "details": {
        "Message.ReservedCapacityFreed.Flag": "true",
        "cpuNumber": "1",
        "cpuSpeed": "512",
        "memory": "512"
      },
      (...)
    }
  },
  "jobresultcode": 0,
  "jobresulttype": "object",
  "jobstatus": 1,
  "userid": "203addbc-4b71-11ef-87b4-cec422422af1"
}
```

Given that, could you check the return of the `queryAsyncJobResult` API?

**Reply by YashasAnand** [reply]
Thank you @bernardodemarco , i tried with cloudmonkey as you suggested, it threw the proper error that it expects cpu speed as well, I was not passing cpu speed in details parameter. Thank you very much for you guidance!
